integrations/access/accesslist/app.go

```diff
@@ -0,0 +1,233 @@
+
+	a.job.SetReady(true)
+	for {
+		select {
+		case <-remindInterval.Next():
+			log.Info("Looking for Access List Review reminders")
+
+			var nextToken string
+			var err error
+			for {
+				var accessLists []*accesslist.AccessList
+				accessLists, nextToken, err = a.apiClient.AccessListClient().ListAccessLists(ctx, 0 /* default page size */, nextToken)
+				if err != nil {
+					log.Errorf("error listing access lists: %v", err)
+					continue
+				}
+
+				for _, accessList := range accessLists {
+					if err := a.notifyForAccessListReviews(ctx, accessList); err != nil {
+						log.WithError(err).Warn("Error notifying for access list reviews")
+					}
+				}
+
+				if nextToken == "" {
+					break
+				}
+			}
+		case <-ctx.Done():
+			log.Info("Access list monitor is finished")
+			return nil
+		}
+	}
+}
+
+// notifyForAccessListReviews will notify if access list review dates are getting close. At the moment, this
+// only supports notifying owners.
+func (a *App) notifyForAccessListReviews(ctx context.Context, accessList *accesslist.AccessList) error {
+	log := logger.Get(ctx)
+	allRecipients := make(map[string]common.Recipient, len(accessList.Spec.Owners))
+
+	now := a.clock.Now()
+	// Find the current notification window.
+	notificationStart := accessList.Spec.Audit.NextAuditDate.Add(-accessList.Spec.Audit.Notifications.Start)
+
+	// If the current time before the notification start time, skip notifications.
+	if now.Before(notificationStart) {
+		log.Debugf("Access list %s is not ready for notifications, notifications start at %s", accessList.GetName(), notificationStart.Format(time.RFC3339))
+		return nil
+	}
+
+	// Get the owners from the bot as recipients.
+	for _, owner := range accessList.Spec.Owners {
+		recipient, err := a.bot.FetchRecipient(ctx, owner.Name)
+		if err != nil {
+			log.Debugf("error getting recipient %s", owner.Name)
+			continue
+		}
+		allRecipients[owner.Name] = *recipient
+	}
+
+	if len(allRecipients) == 0 {
+		return trace.NotFound("no recipients could be fetched for access list %s", accessList.GetName())
+	}
+
+	// Try to create base notification data with a zero notification date. If these objects already
+	// exist, that's okay.
+	userNotifications := map[string]time.Time{}
+	for _, recipient := range allRecipients {
+		userNotifications[recipient.Name] = time.Time{}
+	}
+	_, err := a.pluginData.Create(ctx, accessList.GetName(), pd.AccessListNotificationData{
+		UserNotifications: userNotifications,
+	})
+
+	// Error is okay so long as it's already exists.
+	if err != nil && !trace.IsAlreadyExists(err) {
+		return trace.Wrap(err, "during create")
+	}
+
+	// Calculate weeks from start.
+	weeksFromStart := now.Sub(notificationStart) / oneWeek
+	windowStart := notificationStart.Add(weeksFromStart * oneWeek)
+
+	recipients := []common.Recipient{}
+	_, err = a.pluginData.Update(ctx, accessList.GetName(), func(data pd.AccessListNotificationData) (pd.AccessListNotificationData, error) {
```

Comment:
Shouldn't we update plugin data only after we've successfully sent notifications?